### ISSUE TITLE
Compress files & messages aynchronously or in a separate thread

### ISSUE THREAD

**Issue by piraka9011** [ros2/rosbag2#274]
Original request  from #152:
> Compression is applied to each message asynchronously in parallel by a thread from a thread pool (with command line option to specify number of compression threads).

-----

Compression currently occurs in a single threaded workflow. To avoid potentially dropping messages from the message queue, compression should occur asynchronously or in a separate thread.

**Comment by pjreed** [comment]
I'd be interested in working on this if nobody else is doing so right now.  Thanks!

**Comment by Karsten1987** [comment]
sounds good to me!

**Comment by emersonknapp** [comment]
@pjreed nice! I've assigned the ticket you and moved the issue to In Progress on the ZenHub board.

**Comment by pjreed** [comment]
I've got a working implementation for this, although while reviewing it, I realized there are a couple other things we might want to address.

If the incoming queue overflows, the result is that in message compression mode, the messages will be lost, and in file compression mode, the file will still exist but will not be compressed.  That's not really worse than the current behavior, which will end up dropping messages while the main thread is compressing, but I think it would be preferable for messages & files to be stored uncompressed rather than completely dropped.

The catch here is that the current format seems to assume that if compression is enabled, then *all* messages or files are compressed.  I haven't actually reproduced the condition in order to test it, but I suspect something would crash or otherwise throw an exception when trying to read bags with mixed-compression data.  What do you think a reasonable approach to handling this would be?  We could leave that as-is for now and open another issue to address it, or also have this PR handle playing mixed-compression bags, or if somebody else has another idea...

**Comment by piraka9011** [comment]
> I suspect something would crash or otherwise throw an exception when trying to read bags with mixed-compression data.

Correct, most likely what will happen is that when you compress/decompress an uncompressed message/file, Zstd will throw an error since it's not in the format it expects. (Zstd adds some frame metadata to compressed objects).

I don't know if it's within the scope of rosbag2 to handle mixed compression data (maybe it can be a CLI option...?) but one way could be checking the error type (list [here](https://github.com/facebook/zstd/blob/0ed07f6dfe942627e7724e5b910e23ec3cd8ece8/lib/common/zstd_errors.h#L39)) and if it's a frame/checksum issue then we can try to read the message/file uncompressed.

**Comment by emersonknapp** [comment]
From the conversation in https://github.com/ros2/rosbag2/pull/511 - we're thinking about trying to consolidate this compression writer logic with the regular writer logic. If we took that approach, we could have a single queue-processing mechanism on threads (rather than making it compression-specific)

The thread receiving messages would simply toss them on the queue, then the worker thread would take  care of convert->compress->write. This structure would also help with https://github.com/ros2/rosbag2/issues/436

Then I think there would be no possibility of "mixed-compression". Either the queues can keep up (and the messages _must_ be converted+compressed before being written to storage), or messages start being dropped entirely.

**Comment by pjreed** [comment]
Since #511 looks like it's still under development, do you think it'd be a good idea to go ahead and merge this in for now and then replace it later, or would it be better just to wait for that to be finished?

**Comment by emersonknapp** [comment]
We should go ahead with your change - the time frame on that restructuring is uncertain, we're pulled in a few directions at the moment. Sorry to make you wait on that, I had high hopes heh
